### PR TITLE
[WIP] PCA9547 support

### LINKED ISSUE #1857: support PCA9547 (Kasli-SoC assembly option) in i2c.py
Suggestion: expose the I2C switch functions from the runtime to the kernels, and use them in i2c.py instead of reimplementing them as is currently done. The artiq-zynq runtime already supports PCA9547/PCA9548 autodetection.
Needs support in DRTIO aux as well.

### PR DESCRIPTION
**Pull request by Spaqin**
# ARTIQ Pull Request

## Description of Changes

As #1857 mentions - Kasli SoC may use PCA9547 I2C switch. Suggested solution was to expose the select function (which is already done in firmware) to i2c.py, with a more generic API that would be suited to both available PCA9547 and PCA9548. This PR aims to do that. That includes also exposing the function with DRTIO.

Most importantly, this should fix Kasli-SoC behavior with cards that communicate over I2C.

Not well tested yet - Kasli firmware compiles, but haven't tested on a crate yet - thus marked as WIP - to gather initial feedback.

### Related Issue

Closes #1857 

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [ ] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [ ] Close/update issues.
### Code Changes

- [ ] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [ ] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [x] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/coredevice/i2c.py [hunk 1]
```diff
@@ -152,20 +159,13 @@ def __init__(self, dmgr, busno=0, address=0xe8, core_device="core"):
...
         """Enable one channel.
 
         :param channel: channel number (0-7)
         """
-        self.select(1 << channel)
+        # pca954x select expects address in 7-bit form (adds read/write bit on its own)
+        i2c_pca954x_select(self.busno, self.address >> 1, channel, clear)
 
     @kernel
     def readback(self):
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:171]: I think that doesn't work anymore. Remove.

File: artiq/coredevice/i2c.py [hunk 2]
```diff
@@ -152,20 +159,13 @@ def __init__(self, dmgr, busno=0, address=0xe8, core_device="core"):
...
-
-    @kernel
-    def set(self, channel):
+    def set(self, channel, clear=False):
         """Enable one channel.
 
         :param channel: channel number (0-7)
         """
-        self.select(1 << channel)
+        # pca954x select expects address in 7-bit form (adds read/write bit on its own)
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:167]: pca954x_select()

File: artiq/coredevice/i2c.py [hunk 3]
```diff
@@ -33,6 +33,11 @@ def i2c_read(busno: TInt32, ack: TBool) -> TInt32:
     raise NotImplementedError("syscall not simulated")
 
 
+@syscall(flags={"nounwind", "nowrite"})
+def i2c_pca954x_select(busno: TInt32, address: TInt32, channel: TInt32, clear: TBool) -> TNone:
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:37]: Probably easier to do one-hot decoding in Rust than add a parameter. It's also backward-compatible.

**Comment by sbourdeauducq** [comment]
Needs an update of RELEASE_NOTES  - though with the one-hot parameter, it'll be mostly backward-compatible. Basically explain the change and that only one channel can be selected at most.

File: artiq/coredevice/i2c.py [hunk 4]
```diff
@@ -137,8 +142,10 @@ def i2c_read_many(busno, busaddr, addr, data):
         i2c_stop(busno)
 
 
-class PCA9548:
-    """Driver for the PCA9548 I2C bus switch.
+class PCA954X:
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:145]: Just call it I2CSwitch (and same for the syscall). We might use others in the future.

File: artiq/coredevice/i2c.py [hunk 5]
```diff
@@ -33,6 +33,11 @@ def i2c_read(busno: TInt32, ack: TBool) -> TInt32:
     raise NotImplementedError("syscall not simulated")
 
 
+@syscall(flags={"nounwind", "nowrite"})
+def i2c_pca954x_select(busno: TInt32, address: TInt32, mask: TInt32) -> TNone:
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:37]: rename

File: RELEASE_NOTES.rst [hunk 6]
```diff
@@ -33,6 +33,8 @@ Highlights:
   warning is logged. The warning is additional to the one already printed in the core device log upon
   detection of the error.
 * Removed worker DB warning for writing a dataset that is also in the archive
+* PCA9547 I2C switch support on platforms that may use it. 
+  To keep compatibility between platforms, only 1 channel can be selected with the switch.
```
Comment (sbourdeauducq) [inline RELEASE_NOTES.rst:37]: readback is also removed

File: artiq/firmware/libboard_artiq/si5324.rs [hunk 7]
```diff
@@ -179,17 +179,17 @@ fn init() -> Result<()> {
 
     #[cfg(soc_platform = "kasli")]
     {
-        i2c::pca9548_select(BUSNO, 0x70, 0)?;
-        i2c::pca9548_select(BUSNO, 0x71, 1 << 3)?;
+        i2c::switch_select(BUSNO, 0x70 << 1, 0)?;
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_artiq/si5324.rs:182]: I would keep the 7-bit address here which is what the rest of the firmware does IIRC. Modify the kernel side as appropriate.

File: artiq/firmware/ksupport/nrt_bus.rs [hunk 8]
```diff
@@ -43,6 +43,17 @@ pub mod i2c {
             data
         }) as i32
     }
+
+    pub extern fn switch_select(busno: i32, address: i32, channel: i32, clear: bool) {
```
Comment (sbourdeauducq) [inline artiq/firmware/ksupport/nrt_bus.rs:47]: DId you compile-test this?
Anyway I think the clear/mask argument should be removed?

File: artiq/coredevice/i2c.py [hunk 9]
```diff
@@ -137,8 +142,10 @@ def i2c_read_many(busno, busaddr, addr, data):
         i2c_stop(busno)
 
 
-class PCA9548:
-    """Driver for the PCA9548 I2C bus switch.
+class I2CSwitch:
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:145]: The readback function still isn't going to work. Remove.

File: artiq/coredevice/i2c.py [hunk 10]
```diff
@@ -157,12 +164,11 @@ def select(self, mask):
 
         :param mask: Bit mask of enabled channels
         """
-        i2c_write_byte(self.busno, self.address, mask)
+        i2c_pca954x_select(self.busno, self.address, mask)
```
Comment (sbourdeauducq) [inline artiq/coredevice/i2c.py:167]: You could just remove the ``select`` function, only keep ``set``, and add ``unset``.

File: RELEASE_NOTES.rst (hunk 6 above)
Comment (sbourdeauducq) [reply RELEASE_NOTES.rst:37]: Mention the class rename.

File: artiq/firmware/libboard_misoc/i2c.rs [hunk 11]
```diff
@@ -188,13 +188,16 @@ mod imp {
...
 
-    pub fn pca9548_select(busno: u8, address: u8, channels: u8) -> Result<(), &'static str> {
+    pub fn switch_select(busno: u8, address: u8, mask: u8) -> Result<(), &'static str> {
+        // address in 7-bit form
+        // mask in format of 1 << channel (or 0 for disabling output)
+        // PCA9548 support only for now
         start(busno)?;
         if !write(busno, address << 1)? {
-            return Err("PCA9548 failed to ack write address")
+            return Err("I2C Switch failed to ack write address")
```
Comment (sbourdeauducq) [inline artiq/firmware/libboard_misoc/i2c.rs:197]: We know it's a PCA9548 at this point. This error message does not need to be edited.

**Comment by sbourdeauducq** [comment]
Looks OK except this minor error message detail and release note merge conflict. Please go ahead and test.

**Comment by Spaqin** [comment]
Compile tested with a (random) Kasli standalone and satellite (to cover runtime and satman). Hardware tests with Kasli will have to wait a bit.

**Comment by sbourdeauducq** [comment]
Please fix the merge conflicts and test on KC705 (rpi-1).

**Pull request merged**